package github.com/hashicorp/terraform-provider-aws/internal/service/ec2
import (
	"context"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/ec2"
	awstypes "github.com/aws/aws-sdk-go-v2/service/ec2/types"
	"github.com/hashicorp/aws-sdk-go-base/v2/tfawserr"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
	"github.com/hashicorp/terraform-provider-aws/internal/enum"
	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
)
func waitVPCBlockPublicAccessExclusionDeleted(ctx context.Context, conn *ec2.Client, id string, timeout time.Duration) (*awstypes.VpcBlockPublicAccessExclusion, error) {
	stateConf := &retry.StateChangeConf{
		Pending: enum.Slice(
			awstypes.VpcBlockPublicAccessExclusionStateDeleteInProgress,
			// There might API inconsistencies where even after invoking delete, the Describe might come back with a CreateComplete or UpdateComplete status (the status before delete was invoked).
			// To account for that, we are also adding those two statuses as valid statues to retry.
			awstypes.VpcBlockPublicAccessExclusionStateCreateComplete,
			awstypes.VpcBlockPublicAccessExclusionStateUpdateComplete,
		),
		Target:  []string{},
		Refresh: statusVPCBlockPublicAccessExclusion(ctx, conn, id),
		Timeout: timeout,
	}

	outputRaw, err := stateConf.WaitForStateContext(ctx)

	if output, ok := outputRaw.(*awstypes.VpcBlockPublicAccessExclusion); ok {
		tfresource.SetLastError(err, errors.New(aws.ToString(output.Reason)))

		return output, err
	}

	return nil, err
}
